File: gb/GameBoy/Mappers/MBC1.cs
namespace gbemu.GameBoy.Mappers;

public class MBC1 : IMapper
{
    List<byte[]> romBanks;
    int bank;

    byte[] ram;
    bool ramEnabled;

    public MBC1()
    {
        romBanks = new List<byte[]>();
        bank = 0;
        ramEnabled = false;
        ram = new byte[0x2000];
    }

    public byte[] GetERAM()
    {
        return ram;
    }

    public void LoadROM(byte[] data)
    {
        int totalBanks = data.Length / 0x4000;
        int offset = 0;

        romBanks.Clear();

        for (int i = 0; i < totalBanks; i++)
        {
            romBanks.Add(new byte[0x4000]);
            Array.Copy(data, offset, romBanks[i], 0, 0x4000);
            offset += 0x4000;
        }
    }

    public byte ReadROM(ushort address)
    {
        if (address < 0x4000)
            return romBanks[0][address];
        else if (address < 0x8000)
            return romBanks[bank + 1][address ^ 0x4000];
        else
            return 0;
    }

    public byte ReadRAM(ushort address)
    {
        if (address < 0x2000 && ramEnabled)
            return ram[address];
        else
            return 0xFF;
    }

    public void WriteROM(ushort address, byte data)
    {
        if (address < 0x2000)
        {
            if ((data & 0x0F) == 0x0A)
            {
                ramEnabled = true;
            }
            else
            {
                ramEnabled = false;
            }
        }
        else if (address < 0x4000)
        {
            bank = (data & 0x1F) - 1;
            if (bank < 0)
                bank = 0;
        }
    }

    public void WriteRAM(ushort address, byte data)
    {
        if (address < 0x2000 && ramEnabled)
            ram[address] = data;
    }
}

File: src/gb/mappers/mbc1.c
#include "mbc1.h"

uint8_t gb_mapper_mbc1_read(gb_cart *cart, uint16_t address) {
    if (address < 0x4000) {
        return cart->rom[0][address];
    }
    if (address < 0x8000) {
        if (cart->rom_bank < cart->rom_len)
            return cart->rom[cart->rom_bank][address ^ 0x4000];
    }

    return 0xff;
}

void gb_mapper_mbc1_write(gb_cart *cart, uint16_t address, uint8_t data) {
    if (address >= 0x2000 && address < 0x4000) {
        uint8_t bank = data & 0b00011111;

        if (bank == 0) {
            cart->rom_bank = 0;
        }
        else {
            cart->rom_bank = (bank - 1) % cart->rom_len;
        }
    }
}
